More type safe implementation of BType, cleanups in GenBCode

lrytz opened the pull request:
More type safe `BType` implementation. Various cleanups in the `GenBCode` backend. Also some cleanups and clarifications in `Names`. More details in the individual commit messages.

lrytz committed:
Cleanups for tracing in GenBCode, improve some comments

lrytz committed:
Minor cleanups and commenting around BType.

Use `length` instead of `size` on arrays in `reflect/internal/Names`.

lrytz committed:
Cleanups in Names.scala

Enter only fully constructed Name objects into the hash table, this
should make lookupTypeName thread-safe.

Clean up lookupTypeName - the hash code for a type name is the same
as for its correspondent term name.

Going from a type name toTermName should never create a new TermName
instance. Assert that.

lrytz committed:
More cleanups and commenting in GenBCode

lrytz committed:
Re-use the exsiting range of the name table when using subName.

The sub-name can just point to a smaller range of the array.

lrytz committed:
Move class BType to separate component, include it by composition.

Rewire BType in its new location with the rest of the backend.

Remove the now-empty BCodeGlue.

lrytz committed:
Unit tests for new BType classes

lrytz commented:
Review by @gkossakowski (everybody's feedback is appreciated, of course).

lrytz commented:
So one problem we have is that PR validation doesn't run with `-Ybackend:GenBCode`, we only have a nightly. I'm running a [scala-checkin-manual](https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/1218/parameters/) with the appropriate compiler flags. Is there an urge to automate that? I'm OK to do it manually for my PRs, and that might be enough since the code is under `-Y`.

gkossakowski commented:
I think it's fine to use scala-checkin-manual for now. However, we shouldn't have any failures with regular pr validation right?

lrytz commented:
So the first of the two is expected, it's explained in the commit message. It simplifies reviewing, I'll squash it after (I can also do it now, as you prefer). The second should have built successfully, I'll look into it.

I'll also adding some unit tests for `reflect.internal.Names`.

gkossakowski commented:
Ah, I didn't look at commit messages. I'll start reviewing now.

lrytz committed:
Unit tests for reflect/internal/Names

lrytz commented:
@gkossakowski I fixed the build issue and squashed the refactoring so that all commits build. If you want to review https://github.com/lrytz/scala/commit/70d179084673fa317c4380529c9ab58f42de19e8 in two steps, they are https://github.com/lrytz/scala/commit/476d837a9fca33290b7f3a374c18d9ccab5471ba and https://github.com/lrytz/scala/commit/d185b647a5fd4d4ca15712f1bfdfb57497bef1a2.

lrytz committed:
Rewrite BType to a type hierarchy.

This makes BTypes more type safe.

It fixes at least one bug: the old BTypes are not actually
hash-consed. This worked because the equals method used to have a
fallback that would compare the represented string
character-by-character.

adriaanm commented:
PLS SYNCH

adriaanm commented on src/compiler/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala:
```diff
@@ -582,15 +580,14 @@ abstract class BCodeIdiomatic extends BCodeGlue {
     // can-multi-thread
-    final def checkCast(tk: BType) {
-      assert(tk.isRefOrArrayType, s"checkcast on primitive type: $tk")
+    final def checkCast(tk: RefBType) {
```
procedure syntax was deprecated, btw

retronym replied:
Technically, the past tense only applies in the frame of reference of the terminator that we sent forward in time to deprecate this in the `-Xfuture`. Thoughtfully, the future folk sent a T-1000, currently assuming the form of @samuelgruetter, back to our time to create [srewrite](https://github.com/samuelgruetter/srewrite).

som-snytt replied:
The real challenge is to go back in git history without triggering Armageddon.

scala-jenkins commented:
(kitty-note-to-self: ignore 43931620)
:cat: Synchronaising! :pray:

adriaanm commented on src/compiler/scala/tools/nsc/transform/Mixin.scala:
```diff
@@ -391,7 +391,7 @@ abstract class Mixin extends InfoTransform with ast.TreeDSL {
             sourceModule setPos sym.pos
             if (sourceModule.flags != MODULE) {
-              log("!!! Directly setting sourceModule flags from %s to MODULE".format(sourceModule.flagString))
+              log("!!! Directly setting sourceModule flags for $sourceModule from %s to MODULE".format(sourceModule.flagString))
```
interpolator refactoring interrupted half

som-snytt replied:
This must be under `-Xlint:no-missing-interpolator`.

gkossakowski commented:
LGTM

Good work!

lrytz commented:
addressed feedback: introduced `asClassBType` and friends (https://github.com/lrytz/scala/commit/fcc3842d602114b280e0d9e3dc3adeb5f2f027e7#commitcomment-6416102), fixed a half-hearted string interpolation, removed some procedure syntax.

`-Ybackend:GenBCode` build running here: https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/1220/console

lrytz commented:
PLS REBUILD/pr-scala@4b800a1

scala-jenkins commented:
(kitty-note-to-self: ignore 44785613)
:cat: Roger! Rebuilding pr-scala for 4b800a12. :rotating_light:

lrytz commented:
PLS REBUILD/pr-scala@3914e2b

lrytz commented:
PLS REBUILD/pr-scala@ec3bcc2

scala-jenkins commented:
(kitty-note-to-self: ignore 44787036)
:cat: Roger! Rebuilding pr-scala for 3914e2bc. :rotating_light:

scala-jenkins commented:
(kitty-note-to-self: ignore 44787046)
:cat: Roger! Rebuilding pr-scala for ec3bcc2d. :rotating_light:

lrytz commented:
NOLITTER!

scala-jenkins commented:
(kitty-note-to-self: ignore 44788979)
:cat: cleaning up... sorry! :cat:

som-snytt commented:
I've never seen kitteh apologizz for her bad behavior.  Is she feeling OK?

adriaanm commented:
She's feeling a bit vulnerable these days. Not insecure, though.
